src/os/linux/xdg.rs

```diff
@@ -0,0 +1,127 @@
+pub fn query_mime_entry(mime_type: &str) -> Option<PathBuf> {
+    let out = bossy::Command::impure("xdg-mime")
+        .add_args(&["query", "default", mime_type])
+        .run_and_wait_for_output()
+        .ok()?;
+    if let Ok(out_str) = out.stdout_str() {
+        log::debug!("query_mime_entry got output \"{}\"", out_str);
+        if !out_str.is_empty() {
+            return Some(out_str.trim().into())
+        }
+    }
+    None
+}
+
+// Returns the first entry on that directory whose filename is equal to target.
+//
+// This spec is what makes me believe the search is recursive:
+// https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html
+// This other one does not give that idea:
+// https://specifications.freedesktop.org/menu-spec/latest/ar01s02.html
+pub fn find_entry_in_dir(dir_path: &Path, target: &Path) -> Option<PathBuf> {
+    for entry in dir_path.read_dir().ok()? {
+        if let Ok(entry) = entry {
+            // If it is a file with that same _filename_ (not full path)
+            if entry.path().is_file() && entry.file_name() == target {
+                return Some(entry.path().into())
+            } else if entry.path().is_dir() {
+                // I think if there are any dirs on that directory we have to
+                // recursively search on them
+                if let Some(result) = find_entry_in_dir(&entry.path(), target) {
+                    return Some(result);
+                }
+            }
+        }
+    }
+    None
+}
+
+// Creates a command from the path to a freedesktop "Desktop Entry" file.
+// These kind of files are generally named "something.desktop"
+pub fn command_from_freedesktop_entry(entry: &Path) -> Option<String> {
+    let parsed = parse_entry(entry).ok()?;
+
+    let exec_command = parsed.section("Desktop Entry")
+        .attr("Exec")?
+        .into();
+
+    Some(exec_command)
+}
+
+pub fn find_entry_by_app_name(dir_path: &Path, app_name: &str) -> Option<FreeDesktopEntry> {
+    for entry in dir_path.read_dir().ok()? {
+        if let Ok(entry) = entry {
+            // If it is a file we open it
+            if entry.path().is_file() {
+                if let Ok(parsed) = parse_entry(entry.path()) {
+                    if let Some(name) = parsed.section("Desktop Entry").attr("Name") {
+                        if name == app_name {
+                            return Some(parsed);
+                        }
+                    }
+                }
+            } else if entry.path().is_dir() {
+                // Recursively keep searching if it is a directory
+                if let Some(result) = find_entry_by_app_name(&entry.path(), app_name) {
+                    return Some(result);
+                }
+            }
+        }
+    }
+    None
+}
+
+// The exec field of the FreeDesktop entry may contain some flags that need to
+// be replaced by parameters or even other stuff. The other things are still
+// not implemented
```

Comment:
I found some prior art here: https://github.com/jceb/dex/blob/master/dex#L447

It fortunately looks pretty straightforward, at least as far as parsing state machines go...